pyCalculations/fieldtracer.py

```diff
@@ -190,89 +190,156 @@ def static_field_tracer_3d( vlsvReader, coord_list, max_iterations, dx, directio
+
+   # Trace the field of fg variables
+   def fg_trace(fg):
+      # Create x, y, and z coordinates:
+      xsize = fg.shape[0]
+      ysize = fg.shape[1]
+      zsize = fg.shape[2]
+      xmin = vlsvReader.read_parameter('xmin')
+      xmax = vlsvReader.read_parameter('xmax')
+      ymin = vlsvReader.read_parameter('ymin')
+      ymax = vlsvReader.read_parameter('ymax')
+      zmin = vlsvReader.read_parameter('zmin')
+      zmax = vlsvReader.read_parameter('zmax')
+      sizes = np.array([xsize, ysize, zsize])
+      maxs = np.array([xmax, ymax, zmax])
+      mins = np.array([xmin, ymin, zmin])
+      dcell = (maxs - mins)/(sizes.astype('float'))
+      x = np.arange(mins[0], maxs[0], dcell[0]) + 0.5*dcell[0]
+      y = np.arange(mins[1], maxs[1], dcell[1]) + 0.5*dcell[1]
+      z = np.arange(mins[2], maxs[2], dcell[2]) + 0.5*dcell[2]
+      coordinates = np.array([x,y,z], dtype=object)
+
+      # Create grid interpolation of vector field (V)
+      interpolator_face_V_0 = interpolate.RegularGridInterpolator((x-0.5*dcell[0], y, z), fg[:,:,:,0], bounds_error = False, fill_value = np.nan)
+      interpolator_face_V_1 = interpolate.RegularGridInterpolator((x, y-0.5*dcell[1], z), fg[:,:,:,1], bounds_error = False, fill_value = np.nan)
+      interpolator_face_V_2 = interpolate.RegularGridInterpolator((x, y, z-0.5*dcell[2]), fg[:,:,:,2], bounds_error = False, fill_value = np.nan)
+      interpolators = [interpolator_face_V_0, interpolator_face_V_1, interpolator_face_V_2]
+
+      # Trace vector field lines
+      points_traced = [np.array(coord_list)]              # iteratively append traced trajectories to this list
+      points = points_traced[0]
+      N = len(list(coord_list))
+      V_unit = np.zeros([3, N])
+      for i in range(max_iterations):
+         V_unit[0, :] = interpolators[0](points)
+         V_unit[1, :] = interpolators[1](points)
+         V_unit[2, :] = interpolators[2](points)
+         V_mag = np.linalg.norm(V_unit, axis=(0))
+         V_unit = V_unit / V_mag[np.newaxis,:]
+         new_points = points + multiplier*V_unit.T * dx
+         points = new_points
+         points_traced.append( list(points) )             # list of lists of 3-element arrays
+      return points_traced
+
+   def vg_trace(vg, lower_bound, upper_bound):
+      # Search for the unique coordinates in the given seeds only
+      unique_seed_coords,indices = np.unique(seed_coords, axis = 0, return_inverse = True)    # indice here is to reverse the coords order to initial
+      n_unique_seeds = unique_seed_coords.shape[0]
+      points_traced_unique = np.zeros((n_unique_seeds, max_iterations, 3))
+      
+      def find_unit_vector(vg, coord):
+         val_at_point = vlsvReader.read_interpolated_variable(vg,coord)
+         val_mag = np.linalg.norm(val_at_point, axis = 1, keepdims = True)
+         return val_at_point/val_mag
+
+      # The stop condition for tracing, unit: m
+      def stopping_condition(distances, lower_bound, upper_bound):
```

Comment:
This I would generalize and remove entirely from this scope, so that the function
`def static_field_tracer_3d( vlsvReader, seed_coords, max_iterations, dx, direction='+', grid_var = 'vg_b_vol', lower_bound = 5 * 6371000, upper_bound = 40 * 6371000):`
would rather be
`def static_field_tracer_3d( vlsvReader, seed_coords, max_iterations, dx, direction='+', grid_var = 'vg_b_vol', stopping_condition = default_stopping_condition):`

with a new default function that returns False for each point. Then you do in your own script your own function like:

```
lower_bound = 30e6 # or whatever
upper_bound = 1e9
def my_stop(points):
   distances = np.linalg.norm(points[:,:],axis = 1)
   return (distances <= lower_bound) | (distances >= upper_bound)
```
...